tasty-query/shared/src/main/scala/tastyquery/TypeMatching.scala

```diff
@@ -0,0 +1,310 @@
+         */
+        def covariantDisjoint(tp1: Type, tp2: Type, tparam: TypeParamInfo): Boolean =
+          provablyDisjoint(tp1, tp2) && typeparamCorrespondsToField(tycon1, tparam)
+
+        /* Contravariant case: a value where this type parameter is
+         * instantiated to `Any` belongs to both types.
+         */
+        def contravariantDisjoint(tp1: Type, tp2: Type, tparam: TypeParamInfo): Boolean =
+          false
+
+        /* In the invariant case, we also use a stronger notion of disjointness:
+         * we consider fully instantiated types not equal wrt =:= to be disjoint
+         * (under any context). This is fine because it matches the runtime
+         * semantics of pattern matching. To implement a pattern such as
+         * `case Inv[T] => ...`, one needs a type tag for `T` and the compiler
+         * is used at runtime to check it the scrutinee's type is =:= to `T`.
+         * Note that this is currently a theoretical concern since Dotty
+         * doesn't have type tags, meaning that users cannot write patterns
+         * that do type tests on higher kinded types.
+         */
+        def invariantDisjoint(tp1: Type, tp2: Type, tparam: TypeParamInfo): Boolean =
+          provablyDisjoint(tp1, tp2) /*||
+          !isSameType(tp1, tp2) &&
+          fullyInstantiated(tp1) && // We can only trust a "no" from `isSameType` when
+          fullyInstantiated(tp2)    // both `tp1` and `tp2` are fully instantiated.*/
+
+        tp1.args.lazyZip(tp2.args).lazyZip(tycon1.typeParams).exists { (arg1, arg2, tparam) =>
+          val v = tparam.paramVariance
+          if v.isCovariant then covariantDisjoint(arg1, arg2, tparam)
+          else if v.isContravariant then contravariantDisjoint(arg1, arg2, tparam)
+          else invariantDisjoint(arg1, arg2, tparam)
+        }
+
+      case (tp1: TypeProxy, tp2: TypeProxy) =>
+        def trySuperTp1 = tp1 match
+          case TypeRef.OfClass(_) => false
+          case _                  => provablyDisjoint(tp1.superTypeNormalized, tp2)
+        def trySuperTp2 = tp2 match
+          case TypeRef.OfClass(_) => false
+          case _                  => provablyDisjoint(tp1, tp2.superTypeNormalized)
+        trySuperTp1 || trySuperTp2
+      case (tp1: TypeProxy, _) =>
+        provablyDisjoint(tp1.superTypeNormalized, tp2)
+      case (_, tp2: TypeProxy) =>
+        provablyDisjoint(tp1, tp2.superTypeNormalized)
+
+      case _ =>
+        false
+  end provablyDisjoint
+
+  /** Does `tycon` have a field with type `tparam`? */
+  private def typeparamCorrespondsToField(tycon: Type, tparam: TypeParamInfo)(using Context): Boolean =
+    productSelectorTypes(tycon).exists {
+      case tp: TypeRef => tp.isLocalRef(tparam)
+      case _           => false
+    } /*|| tycon.derivesFrom(defn.PairClass)*/
+  end typeparamCorrespondsToField
+
+  private def productSelectorTypes(tycon: Type)(using Context): Iterator[Type] =
+    Iterator
```

Comment:
questionable, why not look for the case Accessors?

Apparently this is what scala 3 does though